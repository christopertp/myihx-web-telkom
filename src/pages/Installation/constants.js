export const DETAIL_FETCHED = 'Installation/detail-fetched';
export const FAILED = 'Installation/failed';
export const LOADING = 'Installation/loading';
export const NPS_FETCHED = 'Installation/nps-fetched';
export const NPS_RESULT = 'Installation/nps-result';
export const GOOD_TO_KNOW = [
  'Installation will take at least 3 hours. Installation may take longer if there are complications. Make sure that you or a contact person is available during that time.',
  'There may be additional charges once the installation begins. These charges will be confirmed to you beforehand.',
  'If the installation is not successful, we will reschedule or issue a refund.',
];
