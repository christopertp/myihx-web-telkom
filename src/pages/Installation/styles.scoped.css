.footer { text-align: center; }
.footer > h5 {
  color: var(--slate-grey);
  font-size: 1rem;
  font-weight: normal
}

.footer > a {
  color: var(--primary-color);
  font-size: 1.125rem;
  font-weight: bold;
  text-transform: uppercase;
}

.good-to-know { margin-bottom: 2rem; }
.good-to-know > ul { margin-top: 0.5rem; }
.good-to-know > ul > li {
  display: grid;
  grid-template-columns: fit-content(100%) auto;
  grid-template-rows: auto auto;
}

.good-to-know > ul > li > div {
  grid-row: 1/4;
  padding-top: 0.5rem;
  margin-right: 0.5rem;
}

.good-to-know > ul > li > div > div { left: 0 }
.good-to-know > ul > li > p {
  color: var(--text-color);
  font-size: 1rem;
  margin-bottom: 1rem;
}

.modal-404 > img { height: 6.25rem; }
.modal-404 > h3 { margin-bottom: 1rem; }

.package { padding: 1rem 0; }

.package > button {
  font-size: 0.875rem;
  font-weight: bold;
  letter-spacing: 0.0625rem;
  margin: auto;
  padding: 0.25rem 0;
}

.package-article {
  overflow-y: hidden;
  transition: height 0.3s ease-in-out;
  width: 100%;
}

.success-date {
  background-color: #f2f2f2;
  margin-top: 1.25rem;
  padding: 1rem;
  text-align: center;
}

.success-date > h3 {
  font-size: 1.25rem;
  margin-bottom: 0.25rem;
}

.success-date > p {
  color: #66686a;
}

.success-date + a {
  margin: 1.25rem 0 0;
}

@media only screen and (max-width: 48em) {
  .card {
    font-size: 0.9rem;
  }

  .card::before {
    height: 0.8rem;
    width: 0.8rem;
  }
  
  .card.checked::after {
    margin-left: 0.3rem;
    margin-top: 0.3rem;
  }

  .footer > h5 { font-size: 0.75rem; }
  .footer > a { font-size: 0.8125rem; }
  
  .good-to-know > ul > li > div { padding-top: 0.3125rem; }
  .good-to-know > ul > li > p { font-size: 0.75rem; }

  .modal-rating > img { height: 10em; }
  .modal-rating > p { 
    font-size: 0.8rem; 
    line-height: 1.5;
  }
  .modal-rating > div { margin: 2rem 0; }
  .modal-rating > div > figure {
    height: 2.5rem;
    width: 2.5rem;
  }
  .modal-rating > div > h4 { font-size: 1rem; }

  .package > button { font-size: 0.8125rem; }

  .question {
    font-size: 0.65rem;
  }

  .success-date > h3 { font-size: 1rem; }
  .success-date > p { font-size: 0.75rem; }
  .success-date + a { margin: auto 0 1.5rem; }
}
