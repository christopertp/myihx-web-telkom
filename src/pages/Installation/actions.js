import { npsGet, npsPost } from '../../services/nps';
import { installationDetail } from '../../services/order';
import { DETAIL_FETCHED, FAILED, NPS_FETCHED, NPS_RESULT, LOADING } from './constants';

export function fetchInstallationDetail() {
  return async dispatch => {
    dispatch(loadingAction(true, 'detail'));
    try {
      const { data } = await installationDetail();
      const { installationAddress, subscription, schedule } = data;
      const address = installationAddress.description;
      const dataProduct = {
        description: subscription.flag,
        id: subscription.productId,
        mbps: subscription.internetPackage.speed,
        name: subscription.packageName,
        price: subscription.price,
      };
      const { scheduleAttempt, timeSlot } = schedule;

      dispatch(fetchedAction(address, dataProduct, scheduleAttempt, timeSlot, schedule.technician));
    } catch (err) {
      dispatch(failedAction(err.code === 404 ? '404' : 'error', 'detail'));
    }
  };
}

export function fetchNps() {
  const key = 'nps';
  return async dispatch => {
    dispatch(loadingAction(true, key));
    try {
      const { data } = await npsGet();
      dispatch(npsFetchedAction(data));
    } catch (err) {
      dispatch(resultAction(err.message));
    }
  };
}

export function fetchNpsSubmit(feedback, rating) {
  const key = 'submit';
  return async dispatch => {
    dispatch(loadingAction(true, key));
    try {
      await npsPost(feedback);
      dispatch(resultAction(rating > 3 ? 'THANKS' : 'OK'));
      dispatch(fetchNps());
    } catch (err) {
      dispatch(resultAction(err.message));
    }
  };
}

function fetchedAction(address, product, rescheduleAttempt, schedule, technician) {
  return { type: DETAIL_FETCHED, address, product, rescheduleAttempt, schedule, technician };
}

function failedAction(message, key) {
  return { type: FAILED, message, key };
}

function loadingAction(isLoading, key) {
  return { type: LOADING, isLoading, key };
}

function npsFetchedAction(questionnaires) {
  return { type: NPS_FETCHED, questionnaires };
}

function resultAction(npsMessage) {
  return { type: NPS_RESULT, npsMessage };
}
