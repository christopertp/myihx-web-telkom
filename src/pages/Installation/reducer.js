import { DETAIL_FETCHED, FAILED, LOADING, NPS_FETCHED, NPS_RESULT } from './constants';

const initialState = {
  address: '',
  isLoading: {
    d: true,
    n: false,
    s: false,
  },
  message: '',
  npsMessage: '',
  product: '',
  questionnaires: {},
  rescheduleAttempt: 0,
  schedule: '',
  technician: '',
};

export default function reducer(state = initialState, action = {}) {
  const { type, isLoading, address, message, npsMessage, key, product,
    rescheduleAttempt, schedule, technician, questionnaires } = action;

  switch (type) {
    case DETAIL_FETCHED:
      return {
        ...state,
        address,
        isLoading: { ...state.isLoading, d: false },
        product,
        rescheduleAttempt,
        schedule,
        technician,
      };
    case FAILED:
      return {
        ...state,
        message,
        isLoading: { ...state.isLoading, [key.charAt(0)]: false },
      };
    case LOADING:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key.charAt(0)]: isLoading },
      };
    case NPS_FETCHED:
      return {
        ...state,
        isLoading: { ...state.isLoading, n: false },
        questionnaires: parseNpsQuestionnaires(questionnaires),
      };
    case NPS_RESULT:
      return {
        ...state,
        isLoading: { ...state.isLoading, s: false },
        npsMessage,
      };
    default:
      return state;
  }
}

function parseNpsQuestionnaires (val) {
  return val.reduce((questionnaires, quest) => {
    if (quest === null) {
      return questionnaires;
    }

    const { bookReserveId, isRepeat, scheduleType,
      scheduleValue, statements, timeLeft, questionnaireId } = quest;
    const questionnaire = statements.reduce((a, v) => {
      const { statementId, question } = v;
      const { values } = v.answers[0];

      return { ...a, [v.answers[0].label]: { question, values, statementId } };
    }, {});

    return {
      ...questionnaires,
      [quest.code]: {
        bookReserveId,
        isRepeat,
        scheduleType,
        scheduleValue,
        timeLeft,
        questionnaireId,
        ...questionnaire,
      },
    };
  }, {});
}
