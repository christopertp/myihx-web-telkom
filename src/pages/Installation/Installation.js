import React, { createContext, useContext, useEffect, useRef, useState } from 'react';
import moment from 'moment';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import Bullet from '../../components/elements/Bullet';
import Button from '../../components/elements/Button';
import Heading from '../../components/elements/Heading';
import SectionPackage from '../../components/elements/InstallationPackageInfo';
import Modal from '../../components/elements/Modal';
import ModalMaxReschedule from '../../components/elements/ModalMaxReschedule';
import ModalReschedule from '../../components/elements/ModalReschedule';
import ModalTechRating from '../../components/elements/ModalTechRating';
import InstallationSchedule from '../../components/fragments/InstallationSchedule';
import InstallationStatus from '../../components/fragments/InstallationStatus';
import InstallationTechnician from '../../components/fragments/InstallationTechnician';
import WhiteCardPlain from '../../components/layouts/WhiteCardPlain';
import WhiteCardSuccessPlain from '../../components/layouts/WhiteCardSuccessPlain';
import { NpsCodes, NpsContext } from '../../contexts';
import { fetchInstallationDetail, fetchNps, fetchNpsSubmit } from './actions';
import { GOOD_TO_KNOW } from './constants';
import styles from './styles.scoped.css';

const Context = createContext({});

export default function Installation() {
  const history = useHistory();
  const { page } = useParams();

  useEffect(() => {
    (page && page !== 'reschedule-success') && history.replace('/installation');
  }, []);

  const contextValue = {};

  return (
    <Context.Provider value={contextValue}>
      {(page === 'reschedule-success') ? <RescheduleSuccess /> : <Detail />}
    </Context.Provider>
  );
}

export function Detail() {
  const { NPS_TECH_INSTALL } = NpsCodes;
  const { showNps } = useContext(NpsContext);
  const dispatch = useDispatch();
  const { isLoading, message, npsMessage, questionnaires, rescheduleAttempt,
    technician } = useSelector(s => s.installation);
  const [modal, setModal] = useState('');
  const code = 'NPS-TECH-RATING';
  const maxReschedule = rescheduleAttempt >= 2 ? 'max-reschedule' : 'reschedule';

  useEffect(() => {
    dispatch(fetchInstallationDetail());
    dispatch(fetchNps());
  }, []);

  useEffect(() => {
    if (message === '404') {
      setModal('404');
    } else if (['OK', 'THANKS'].includes(npsMessage)) {
      setModal('');
      showNps(NPS_TECH_INSTALL);
    }
  }, [message,npsMessage]);

  const onSubmit = (feedback, rating) => {
    dispatch(fetchNpsSubmit(feedback, rating));
  };

  return (
    <>
      <WhiteCardPlain title="Appointment Detail">
        <InstallationSchedule onClick={() => setModal(maxReschedule)} />
        <Package />
        {technician ? (
          <section>
            <InstallationTechnician />
            <InstallationStatus />
          </section>
        ) : ''}
        <GoodToKnow />
        <footer className={styles.footer}>
          <h5>Need Help?</h5>
          <a href="tel:147">Contact Us</a>
        </footer>
      </WhiteCardPlain>
      <ModalReschedule onClose={() => setModal('')} open={modal === 'reschedule'} />
      <ModalMaxReschedule onClose={() => setModal('')} open={modal === 'max-reschedule'} />
      <Modal className={styles['modal-404']} open={modal === '404'}>
        <img alt="" src="/assets/grfx_warning.svg" />
        <h3>No installation found</h3>
        <Button to="/">Back to Home</Button>
      </Modal>
      <ModalTechRating
        code={code}
        isLoading={isLoading}
        message={npsMessage}
        onClose={() => setModal('')}
        onSubmit={onSubmit}
        open={modal === 'rating'}
        questionnaires={questionnaires[code]}
        technicianInfo={{ technician }}
      />
    </>
  );
}

export function Package() {
  const { address, isLoading, product, technician } = useSelector(s => s.installation);
  const [open, setOpen] = useState(false);
  const ref = useRef(null);

  const getHeight = c => (
    [...c].reduce((a, i) => {
      const style = window.getComputedStyle(i);
      const marginTop = style.getPropertyValue('margin-top').split('px')[0];
      const marginBottom = style.getPropertyValue('margin-bottom').split('px')[0];
      return a + i.offsetHeight + (marginBottom / 1) + (marginTop / 1);
    }, 0)
  );

  const height = (ref.current && ref.current.children) ? getHeight(ref.current.children) / 1 : 0;

  const onClickOpen = () => {
    setOpen(!open);
  };

  if (!technician) {
    return (
      <section className={styles.package}>
        <SectionPackage address={address} isLoading={isLoading.d} product={product} />
      </section>
    );
  } else {
    return (
      <section className={styles.package}>
        <article className={styles['package-article']} ref={ref} style={{ height: open ? height : 0 }}>
          <SectionPackage address={address} isLoading={isLoading.d} product={product} />
        </article>
        <Button onClick={onClickOpen} variant="text">
          {open ? 'Hide' : 'View'} Installation Detail
        </Button>
      </section>
    );
  }
}

export function RescheduleSuccess() {
  const { newDate, slot } = useSelector(s => s.rescheduleInstallation);
  const title = 'Reschedule Success!';
  const subtitle = 'Your new installation schedule is';
  return (
    <WhiteCardSuccessPlain icon="success" subtitle={subtitle} title={title}>
      <div className={styles['success-date']}>
        <h3>{moment(newDate).format('dddd, D MMMM YYYY')}</h3>
        <p>Estimated arrival time {slot}</p>
      </div>
      <Button to="/">Back to Home</Button>
    </WhiteCardSuccessPlain>
  );
}

export function GoodToKnow() {
  return (
    <section className={styles['good-to-know']}>
      <Heading>Good to know</Heading>
      <ul>
        {GOOD_TO_KNOW.map((item, idx) => (
          <li key={idx}>
            <Bullet color={['#ee3124', '']} size="0.5rem" />
            <p>{item}</p>
          </li>
        ))}
      </ul>
    </section>
  );
}
