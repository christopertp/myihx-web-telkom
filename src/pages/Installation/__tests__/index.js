import React, { useState, useRef } from 'react';
import { useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import Installation, { Detail, GoodToKnow, Package, RescheduleSuccess } from '../Installation';

jest.mock('../actions', () => ({
  fetchInstallationDetail: jest.fn(),
  fetchNps: jest.fn(),
}));

useSelector.mockImplementation(fn => {
  fn({});
  return { isLoading: {
    d:false,
    n: false,
    s: false,
  } };
});

describe('src/pages/Installation', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Installation />);
    expect(tree).toMatchSnapshot();
  });

  test('render subroute', () => {
    const replace = jest.fn();
    useHistory.mockImplementationOnce(() => ({ replace }));
    useParams.mockImplementationOnce(() => ({ page: 'anything' }));

    const result = Installation();
    expect(replace).toHaveBeenCalledWith('/installation');
    expect(result.props.children.type).toBe(Detail);
  });

  test('render subroute reschedule success', () => {
    useParams.mockImplementationOnce(() => ({ page: 'reschedule-success' }));

    const result = Installation();

    expect(result.props.children.type).toBe(RescheduleSuccess);
  });

  test('Detail', () => {
    useSelector.mockImplementationOnce(() => ({
      isLoading: { n: false },
      questionnaires: { 'NPS-TECH-RATING': {} },
    }));

    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Detail />);
    expect(tree).toMatchSnapshot();
  });

  test('Detail reschedule & error modal', () => {
    const setModal = jest.fn();

    useSelector.mockImplementationOnce(() => ({
      isLoading: {},
      message: '404',
      questionnaires: { 'NPS-TECH-RATING': {} },
      technician: 'test'
    }));
    useState.mockImplementationOnce(() => ['', setModal]);

    const result = Detail();
    expect(setModal).toHaveBeenCalledWith('404');

    result.props.children[0].props.children[0].props.onClick();
    expect(setModal).toHaveBeenCalledWith('reschedule');

    result.props.children[1].props.onClose();
    result.props.children[2].props.onClose();
    expect(setModal).toHaveBeenCalledWith('');
  });

  test('Detail max reschedule', () => {
    const setModal = jest.fn();

    useSelector.mockImplementationOnce(() => ({
      isLoading: { s: false },
      questionnaires: { 'NPS-TECH-RATING': {} },
      rescheduleAttempt: 2 }));
    useState.mockImplementationOnce(() => ['', setModal]);

    const result = Detail();

    result.props.children[0].props.children[0].props.onClick();
    expect(setModal).toHaveBeenCalledWith('max-reschedule');
  });

  test('GoodToKnow', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<GoodToKnow />);
    expect(tree).toMatchSnapshot();
  });

  test('Package', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Package />);
    expect(tree).toMatchSnapshot();
  });

  test('Package Open', () => {
    const setOpen = jest.fn();
    delete global.getComputedStyle;
    global.getComputedStyle = () => ({
      getPropertyValue: () => '10px',
    });

    const children = [{ offsetHeight: 10 }];
    const current = { children };

    useRef.mockImplementationOnce(() => ({ current }));
    useState.mockImplementationOnce(() => [true, setOpen]);
    useSelector.mockImplementationOnce(() => ({
      isLoading: { d: false },
      technician: 'test'
    }));

    const result = Package();
    expect(result.props.children[0].props.style.height).toBe(30);

    result.props.children[1].props.onClick();
    expect(setOpen).toHaveBeenCalled();
  });

  test('RescheduleSuccess', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<RescheduleSuccess />);
    expect(tree).toMatchSnapshot();
  });
});
