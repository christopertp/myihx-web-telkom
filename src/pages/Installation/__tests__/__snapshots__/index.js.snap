// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/Installation Detail 1`] = `
<React.Fragment>
  <WhiteCardPlain
    back=""
    close=""
    help={false}
    icon=""
    size="large"
    subtitle=""
    title="Appointment Detail"
  >
    <InstallationSchedule
      onClick={[Function]}
    />
    <Package />
    
    <GoodToKnow />
    <footer>
      <h5>
        Need Help?
      </h5>
      <a
        href="tel:147"
      >
        Contact Us
      </a>
    </footer>
  </WhiteCardPlain>
  <ModalReschedule
    onClose={[Function]}
    open={false}
  />
  <ModalMaxReschedule
    onClose={[Function]}
    open={false}
  />
  <Modal
    className=""
    onClose={[Function]}
    open={false}
  >
    <img
      alt=""
      src="/assets/grfx_warning.svg"
    />
    <h3>
      No installation found
    </h3>
    <Button
      className=""
      disabled={false}
      onClick={null}
      to="/"
      type="button"
      variant="red"
    >
      Back to Home
    </Button>
  </Modal>
  <ModalTechRating
    code="NPS-TECH-RATING"
    isLoading={
      Object {
        "n": false,
      }
    }
    message=""
    onClose={[Function]}
    onSubmit={[Function]}
    open={false}
    questionnaires={Object {}}
    technicianInfo={
      Object {
        "technician": undefined,
      }
    }
  />
</React.Fragment>
`;

exports[`src/pages/Installation GoodToKnow 1`] = `
<section>
  <Heading
    className=""
  >
    Good to know
  </Heading>
  <ul>
    <li>
      <Bullet
        className=""
        color={
          Array [
            "#ee3124",
            "",
          ]
        }
        connect={false}
        size="0.5rem"
      />
      <p>
        Installation will take at least 3 hours. Installation may take longer if there are complications. Make sure that you or a contact person is available during that time.
      </p>
    </li>
    <li>
      <Bullet
        className=""
        color={
          Array [
            "#ee3124",
            "",
          ]
        }
        connect={false}
        size="0.5rem"
      />
      <p>
        There may be additional charges once the installation begins. These charges will be confirmed to you beforehand.
      </p>
    </li>
    <li>
      <Bullet
        className=""
        color={
          Array [
            "#ee3124",
            "",
          ]
        }
        connect={false}
        size="0.5rem"
      />
      <p>
        If the installation is not successful, we will reschedule or issue a refund.
      </p>
    </li>
  </ul>
</section>
`;

exports[`src/pages/Installation Package 1`] = `
<section>
  <InstallationPackageInfo
    address=""
    isLoading={false}
    product=""
  />
</section>
`;

exports[`src/pages/Installation RescheduleSuccess 1`] = `
<WhiteCardSuccessPlain
  close=""
  icon="success"
  subtitle="Your new installation schedule is"
  title="Reschedule Success!"
>
  <div>
    <h3>
      dddd, D MMMM YYYY
    </h3>
    <p>
      Estimated arrival time 
    </p>
  </div>
  <Button
    className=""
    disabled={false}
    onClick={null}
    to="/"
    type="button"
    variant="red"
  >
    Back to Home
  </Button>
</WhiteCardSuccessPlain>
`;

exports[`src/pages/Installation render 1`] = `
<Provider
  value={Object {}}
>
  <Detail />
</Provider>
`;
